[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively

salilsdesai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87432
* #87431

With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

salilsdesai committed:
[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively

With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

digantdesai reviewed (approved):
LGTM

albanD commented on test/test_public_bindings.py:
```diff
@@ -261,7 +261,7 @@ def test_no_new_bindings(self):
             "unify_type_list",
             "vitals_enabled",
-
+            "VULKAN_AUTOMATIC_GPU_TRANSFER",
```
Do you actually want this to be public API as `torch.VULKAN_AUTOMATIC_GPU_TRANSFER` ?
Is it properly documented as well?

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai committed:
Update on "[Vulkan] Add Vulkan Rewrite to Transfer Inputs and Outputs to Vulkan and CPU Backends Respectively"


With this change, we don't have to manually invoke transferring input and output backends when we run vulkan models.

Graph rewrite code based off of:
- https://github.com/pytorch/pytorch/commit/32efff45ba77f2bb4b1e709613b99070f119745a#diff-a473bddb458dc24225866a45092d6eca064eddd256245d93020e48e216eee4d5R160-R179

Differential Revision: [D39519168](https://our.internmc.facebook.com/intern/diff/D39519168/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39519168/)!

[ghstack-poisoned]

salilsdesai commented:
For sample model
```
class TestModel(torch.nn.Module):
    def forward(self, x):
        return x + x
```
Before Transfer
```
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : Tensor = aten::add(%x.1, %x.1, %2)
  return (%3)
```
After Transfer
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %4 : str = prim::Constant[value="vulkan"]()
  %5 : str = prim::Constant[value="vulkan"]()
  %6 : NoneType = prim::Constant()
  %7 : NoneType = prim::Constant()
  %8 : NoneType = prim::Constant()
  %9 : NoneType = prim::Constant()
  %10 : bool = prim::Constant[value=0]()
  %11 : bool = prim::Constant[value=0]()
  %12 : NoneType = prim::Constant()
  %13 : str = prim::Constant[value="vulkan"]()
  %14 : str = prim::Constant[value="vulkan"]()
  %15 : str = prim::Constant[value="vulkan"]()
  %16 : str = prim::Constant[value="vulkan"]()
  %17 : str = prim::Constant[value="vulkan"]()
  %18 : Device = aten::device(%17)
  %19 : str = prim::Constant[value="vulkan"]()
  %20 : NoneType = prim::Constant()
  %21 : NoneType = prim::Constant()
  %22 : NoneType = prim::Constant()
  %23 : NoneType = prim::Constant()
  %24 : bool = prim::Constant[value=0]()
  %25 : bool = prim::Constant[value=0]()
  %26 : NoneType = prim::Constant()
  %27 : str = prim::Constant[value="vulkan"]()
  %28 : str = prim::Constant[value="vulkan"]()
  %29 : Device = aten::device(%28)
  %30 : NoneType = prim::Constant()
  %31 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %33 : Tensor = aten::to(%x.1, %29, %30, %31, %32)
  %2 : int = prim::Constant[value=1]()
  %3 : Tensor = aten::add(%33, %33, %2)
  %34 : str = prim::Constant[value="cpu"]()
  %35 : str = prim::Constant[value="cpu"]()
  %36 : NoneType = prim::Constant()
  %37 : NoneType = prim::Constant()
  %38 : NoneType = prim::Constant()
  %39 : NoneType = prim::Constant()
  %40 : bool = prim::Constant[value=0]()
  %41 : bool = prim::Constant[value=0]()
  %42 : NoneType = prim::Constant()
  %43 : str = prim::Constant[value="cpu"]()
  %44 : str = prim::Constant[value="cpu"]()
  %45 : str = prim::Constant[value="cpu"]()
  %46 : str = prim::Constant[value="cpu"]()
  %47 : str = prim::Constant[value="cpu"]()
  %48 : Device = aten::device(%47)
  %49 : str = prim::Constant[value="cpu"]()
  %50 : NoneType = prim::Constant()
  %51 : NoneType = prim::Constant()
  %52 : NoneType = prim::Constant()
  %53 : NoneType = prim::Constant()
  %54 : bool = prim::Constant[value=0]()
  %55 : bool = prim::Constant[value=0]()
  %56 : NoneType = prim::Constant()
  %57 : str = prim::Constant[value="cpu"]()
  %58 : str = prim::Constant[value="cpu"]()
  %59 : Device = aten::device(%58)
  %60 : NoneType = prim::Constant()
  %61 : bool = prim::Constant[value=0]()
  %62 : bool = prim::Constant[value=0]()
  %63 : Tensor = aten::to(%3, %59, %60, %61, %62)
  return (%63)
```
After final dead code elimination
```
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %65 : Device = prim::Constant[value="cpu"]()
  %2 : int = prim::Constant[value=1]()
  %31 : bool = prim::Constant[value=0]()
  %30 : NoneType = prim::Constant()
  %64 : Device = prim::Constant[value="vulkan"]()
  %33 : Tensor = aten::to(%x.1, %64, %30, %31, %31)
  %3 : Tensor = aten::add(%33, %33, %2)
  %63 : Tensor = aten::to(%3, %65, %30, %31, %31)
  return (%63)
```

salilsdesai commented:
For sample model
```
class TestModel(torch.nn.Module):
    def forward(self, x):
        return x + x
```
Before Transfer
```
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : Tensor = aten::add(%x.1, %x.1, %2)
  return (%3)
```
After Transfer
```
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %4 : str = prim::Constant[value="vulkan"]()
  %5 : str = prim::Constant[value="vulkan"]()
  %6 : NoneType = prim::Constant()
  %7 : NoneType = prim::Constant()
  %8 : NoneType = prim::Constant()
  %9 : NoneType = prim::Constant()
  %10 : bool = prim::Constant[value=0]()
  %11 : bool = prim::Constant[value=0]()
  %12 : NoneType = prim::Constant()
  %13 : str = prim::Constant[value="vulkan"]()
  %14 : str = prim::Constant[value="vulkan"]()
  %15 : str = prim::Constant[value="vulkan"]()
  %16 : str = prim::Constant[value="vulkan"]()
  %17 : str = prim::Constant[value="vulkan"]()
  %18 : Device = aten::device(%17)
  %19 : str = prim::Constant[value="vulkan"]()
  %20 : NoneType = prim::Constant()
  %21 : NoneType = prim::Constant()
  %22 : NoneType = prim::Constant()
  %23 : NoneType = prim::Constant()
  %24 : bool = prim::Constant[value=0]()
  %25 : bool = prim::Constant[value=0]()
  %26 : NoneType = prim::Constant()
  %27 : str = prim::Constant[value="vulkan"]()
  %28 : str = prim::Constant[value="vulkan"]()
  %29 : Device = aten::device(%28)
  %30 : NoneType = prim::Constant()
  %31 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %33 : Tensor = aten::to(%x.1, %29, %30, %31, %32)
  %2 : int = prim::Constant[value=1]()
  %3 : Tensor = aten::add(%33, %33, %2)
  %34 : str = prim::Constant[value="cpu"]()
  %35 : str = prim::Constant[value="cpu"]()
  %36 : NoneType = prim::Constant()
  %37 : NoneType = prim::Constant()
  %38 : NoneType = prim::Constant()
  %39 : NoneType = prim::Constant()
  %40 : bool = prim::Constant[value=0]()
  %41 : bool = prim::Constant[value=0]()
  %42 : NoneType = prim::Constant()
  %43 : str = prim::Constant[value="cpu"]()
  %44 : str = prim::Constant[value="cpu"]()
  %45 : str = prim::Constant[value="cpu"]()
  %46 : str = prim::Constant[value="cpu"]()
  %47 : str = prim::Constant[value="cpu"]()
  %48 : Device = aten::device(%47)
  %49 : str = prim::Constant[value="cpu"]()
  %50 : NoneType = prim::Constant()
  %51 : NoneType = prim::Constant()
  %52 : NoneType = prim::Constant()
  %53 : NoneType = prim::Constant()
  %54 : bool = prim::Constant[value=0]()
  %55 : bool = prim::Constant[value=0]()
  %56 : NoneType = prim::Constant()
  %57 : str = prim::Constant[value="cpu"]()
  %58 : str = prim::Constant[value="cpu"]()
  %59 : Device = aten::device(%58)
  %60 : NoneType = prim::Constant()
  %61 : bool = prim::Constant[value=0]()
  %62 : bool = prim::Constant[value=0]()
  %63 : Tensor = aten::to(%3, %59, %60, %61, %62)
  return (%63)
```
After final dead code elimination
```
graph(%self : __torch__.___torch_mangle_2.TestModel,
      %x.1 : Tensor):
  %65 : Device = prim::Constant[value="cpu"]()
  %2 : int = prim::Constant[value=1]()
  %31 : bool = prim::Constant[value=0]()
  %30 : NoneType = prim::Constant()
  %64 : Device = prim::Constant[value="vulkan"]()
  %33 : Tensor = aten::to(%x.1, %64, %30, %31, %31)
  %3 : Tensor = aten::add(%33, %33, %2)
  %63 : Tensor = aten::to(%3, %65, %30, %31, %31)
  return (%63)
```
